### ISSUE TITLE
overly restrictive closure borrow-checking

### ISSUE THREAD

**Issue by aliemjay** [rust-lang/rust#104477]
Both of these test cases should pass borrowck: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2015&gist=50a263a9626cb0b08f5cee334df8a852
```rust
mod case1 {
    struct MyTy<'x, 'a, 'b>(std::cell::Cell<(&'x &'a u8, &'x &'b u8)>);
    fn wf<T>(_: T) {}
    fn test<'a, 'b>() {
        |_: &'a u8, x: MyTy<'_, 'a, 'b>| wf(x); // FAIL!
        |x: MyTy<'_, 'a, 'b>, _: &'a u8| wf(x); // PASS!!!
    }
}

mod case2 {
    struct MyTy<'a, 'b, 'x>(std::cell::Cell<(&'a &'x u8, &'b &'x u8)>);
    fn wf<T>(_: T) {}
    fn test<'a, 'b>() {
        |x: MyTy<'a, 'b, '_>| wf(x); // FAIL!
    }
}
```

They both fail because of the unnecessarily restrictive algorithm in [try_propagate_universal_region_error](https://github.com/rust-lang/rust/blob/79146baa9c7e00b716cc41f8660a56cd2acdf8bd/compiler/rustc_borrowck/src/region_infer/mod.rs#L1629):

- `case1` fails because `'_` has multiple "non-local **upper** bounds" `['a, 'b]` and we do propagate the error to **both** of them while choosing a single one would be sufficient.
- `case2` fails because `'_` has multiple "non-local **lower** bounds" `['a, 'b]` and the call to [non_lcal_lower_bound](https://github.com/rust-lang/rust/blob/79146baa9c7e00b716cc41f8660a56cd2acdf8bd/compiler/rustc_borrowck/src/region_infer/mod.rs#L1639) simply returns `None` in this case!

@rustbot label C-bug T-types A-NLL NLL-complete A-borrow-checker

**Comment by Rageking8** [comment]
I have absolutely 0 clue how the code works but is this (case 1) supposed to compile: [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=2dba30bdd448744820ab3489e5364bb6)
```rust
mod case1 {
struct MyTy<'x, 'a, 'b>(std::cell::Cell<(&'x (), &'a u8, &'x &'b u8)>); // Diff here, extra parens
fn wf<T>(_: T) {}
fn test<'a, 'b>() {
|_: &'a u8, x: MyTy<'_, 'a, 'b>| wf(x);
|x: MyTy<'_, 'a, 'b>, _: &'a u8| wf(x);
}}
fn main() {}
```
Found using a custom DIY Fuzzer.
Just wanted to ensure that *if* this is unintended, it can be handled.
If the successful compilation is intended, do kindly ignore this comment. Thanks.

**Comment by compiler-errors** [comment]
@Rageking8 I don't think that the example you produced is related to the issue that @aliemjay is highlighting here, since the changed code no longer requires that means `'x` is related to `'a` in `MyTy`.

(Also, I don't recommend fuzzing random code in issue reports without understanding how it works first, especially if you're looking for changes in pass/fail conditions, and not just trying to find ICEs, for example)

**Comment by lcnr** [comment]
This issue also affects type tests (and i expect to this error even if region outlives were fixed)
```rust
fn reg_outlives<'a, 'b: 'a>() {}
fn type_outlives<'a, T: 'a>() {}

struct Constrain<'a: 'c, 'b: 'c, 'c: 'a>(*mut (&'a (), &'b (), &'c ()));

impl<'a, 'b, 'c> Constrain<'a, 'b, 'c> {
    fn req(&self)
    where
        'c: 'a,
    {
    }
}

fn promote_type_test<'a, 'b: 'a, T>(x: &'a T) {
    |x: Constrain<'a, 'b, '_>| {
        x.req(); // 'c: 'a
        type_outlives::<'a, T>(); // we've got `T: 'a`
    };
}
```
https://github.com/rust-lang/rust/blob/2776bdfe423c9fdfcd6313d678f0852ea26f1309/compiler/rustc_borrowck/src/region_infer/mod.rs#L1012-L1016

There's no reason to look at all lower bounds of our region in case that region is already external.

**Comment by aliemjay** [comment]
@lcnr Your issue can be reproduced and fixed independently using this alternative example:
```rust
#![feature(closure_lifetime_binder)]

fn type_outlives<'a, T: 'a>() {}

struct Constrain<'a: 'c, 'b: 'c, 'c: 'a>(*mut (&'a (), &'b (), &'c ()));

fn promote_type_test<'a, 'b, T>(_: &'a T) {
    for<'x> |_: Constrain<'a, 'b, 'x>| -> () {
        type_outlives::<'x, T>(); //~ `T` may not live long enough
        // unnecessarily requires T: 'b
    };
}
```



> There's no reason to look at all lower bounds of our region in case that region is already external.

I think the way to fix it is a bit more subtle. Consider the example above where `'x` is not external. The real problem is that `non_local_upper_bounds('x)` incorrectly returns `['a, 'b]` when it should return `['a]` only.
https://github.com/rust-lang/rust/blob/2776bdfe423c9fdfcd6313d678f0852ea26f1309/compiler/rustc_borrowck/src/region_infer/mod.rs#L1019
